samples/Grover/Grover.qs

```diff
@@ -0,0 +1,77 @@
+﻿namespace GroverSample {
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Convert;
+    open Microsoft.Quantum.Math;
+    open Microsoft.Quantum.Arrays;
+    open Microsoft.Quantum.Measurement;
+    open Microsoft.Quantum.Intrinsic;
+    open Microsoft.Quantum.Diagnostics;
+    open Microsoft.Quantum.Arithmetic;
+
+    open Qram;
+
+    /// # Summary
+    /// This sample is an adaptation of the Grover sample in the QDK documentation
+    /// that uses a BucketBrigadeQRAM as an oracle, rather than the usual reflection
+    /// about marked states.
+    @EntryPoint()
+    operation GroverSearch(addressSize : Int, markedElements : Int[]) : Int {
+        // First, set up a qRAM with marked elements set to 1
+        let nMarkedElements = Length(markedElements);
+        mutable groverMemoryContents = new MemoryCell[0];
+        for (markedElement in markedElements) {
+            set groverMemoryContents += [MemoryCell(markedElement, [true])];
+		}
+        Message($"Num marked: {nMarkedElements}, Num Iterations: {NIterations(nMarkedElements, addressSize)}");
+
+        // Is there a better way to do this in the situation where the memory register
+        // is not explicitly owned by the qRAM?
+        // Also, the target is not really used here except to execute the query and perform
+        // phase kickback
+        using ((groverQubits, targetQubit, flatMemoryRegister) = 
+            (Qubit[addressSize], Qubit[1], Qubit[2^addressSize])
+        ) {
+                // Create a memory
+                // This part feels really clunky
+                let memoryRegister = MemoryRegister(Most(Partitioned(ConstantArray(2^addressSize, 1), flatMemoryRegister)));
+                let memory = BucketBrigadeQRAMOracle(groverMemoryContents, memoryRegister);
+
+                //DumpRegister("memory.txt", flatMemoryRegister);
+                // Initialize a uniform superposition over all possible inputs.
+                PrepareUniform(groverQubits);
+
+                // Grover iterations - the reflection about the marked element is implemented
+                // as a QRAM phase query. Only the memory cells storing a 1 will produce a phase
+                for (idxIteration in 0..NIterations(nMarkedElements, addressSize) - 1) {
+                    //DumpRegister((), flatMemoryRegister);
```

Comment:
Sorry, I left all these debug statements in here and could have cleaned them up. However, one thing I'll point out for future reference is that having the `_{idxIteration}` for all register dumps makes it easy to see any single iteration with a `cat *_0.txt` or `cat *_1.txt`

Reply:
Great tip!